CI: Gentoo pipelines

Arniiiii opened the pull request:
Related links: 

#16 
https://bugs.gentoo.org/955934
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=120273
https://github.com/Arniiiii/ex_repo/tree/master/dev-cpp/corral

Arniiiii committed:
gentoo pipelines

Arniiiii commented:
I guess it would be preferable before merging these pipelines to think about a possible caching of gentoo image with dependencies and maybe create it new once in a month, because now, with all possible parallelizations and gentoo's binary packages it takes from 6 to 14 minutes for the CI pipelines.

Also the ICE with g++-14 . IDK what to do with it right now, since it's already at gcc's bugzilla.  Maybe try temporary use g++ version in "testing" ( https://packages.gentoo.org/packages/sys-devel/gcc yellow squares correspond to "testing")

dprokoptsev commented:
> I guess it would be preferable before merging these pipelines to think about a possible caching of gentoo image with dependencies and maybe create it new once in a month.

Maybe, but I have my reservations about random docker images, with having to deal with security questions, rate limiting, and whatnot. I think 15 minutes for a CI run is not terribly bad, especially that people are more than welcome to run CI locally (where they can utilize local caching of images with little if any effort). If this becomes too much of a hassle, we can reconsider later.

(FWIW, `msvc-wine` is not blazingly fast either, and I don't think it can be dockerized due to license reasons (but don't quote me on that)).

> Maybe try temporary use g++ version in "testing" ( https://packages.gentoo.org/packages/sys-devel/gcc yellow squares correspond to "testing".

This sounds good -- after all, that version did compile the test suite without ICEing.
Also I guess it might make sense to nail down a particular gcc version, so any further updates won't silently sneak into the CI pipeline and cause any hard-to-debug regressions.

Arniiiii commented:
> > Maybe try temporary use g++ version in "testing" ( https://packages.gentoo.org/packages/sys-devel/gcc yellow squares correspond to "testing".
> 
> This sounds good -- after all, that version did compile the test suite without ICEing. Also I guess it might make sense to nail down a particular gcc version, so any further updates won't silently sneak into the CI pipeline and cause any hard-to-debug regressions.

Then CI will download_sources-unpack-build-install gcc, not download_binary_pkg-install , so it will compile gcc, which can be +1 hour on a less powerful machines, like some of github's runners. 

This can be fixed in next ways:
  a. Somehow cache whole docker image
  b. Somehow have a gentoo container running that already compiled gcc with testing version and make host of binaries out of it, which is easy, but requires to have such server running.
  c. Somehow cache as artifact a binary package of a version of compiled gcc in a gentoo docker, then place it to /var/cache/binpkg/sys-devel/gcc-${VERSION} and then install it. 
  
Problems:
  a. I don't have even slightest idea how to do this
  b. It means to have someone running a server.
  c. Just problem of knowing correct syntax of github's CI yaml files.

dprokoptsev commented:
I thought gentoo did provide pre-built packages? If not, then indeed building gcc as a part of a corral CI run does not sound like a good idea to me.

If your concern is that gentoo may have compiler versions different from what corral CI runs on, what can be done instead is using docker images available on https://hub.docker.com/_/gcc to run CI on all released gcc versions since gcc-12 (idk if there's anything similar for clangs, though).

Arniiiii commented:
> I thought gentoo did provide pre-built packages? 

Gentoo provides pre-built packages for stable versions for some popular USE flags combinations. Also it's relatively easy to make your own pre-built packages and share them.

Arniiiii commented:
It appears I was wrong: there are binary packages for all versions that are testing : https://quantum-mirror.hu/mirrors/pub/gentoo/releases/amd64/binpackages/23.0/x86-64/sys-devel/gcc/

Arniiiii committed:
make gentoo gcc ci use latest gcc in testing

in a nutshell:
for gentoo gcc pipeline:
 1. use latest testing gcc and get it as package managers
 2. change default (the `cc` and `c++` ) gcc version to latest available
 3. update for sure environmental variable: `source /etc/profile`
for clang profile add `source /etc/profile`.

Arniiiii committed:
WIP: test latest testing gcc-14

Arniiiii commented:
Interesting.

latest testing g++-14 ( `c++ (Gentoo 14.2.1_p20250419 p8) 14.2.1 20250419` , the version you said you tested at https://github.com/hudson-trading/corral/pull/16#issuecomment-2887089374 )  : 
https://github.com/hudson-trading/corral/actions/runs/15147855267/job/42587807526?pr=17#step:5:11538
```
/__w/corral/corral/test/basic_test.cc:1011:15: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
 1011 |             (*reinterpret_cast<Fn*>(&obj))(args...);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

latest testing g++-15 (`sys-devel/gcc-15.1.0`):
https://github.com/hudson-trading/corral/actions/runs/15147855267/job/42587808555?pr=17#step:5:2825
```
g++: internal compiler error: Segmentation fault signal terminated program cc1plus
```

latest stable g++-14 : my first ICE

eli-schwartz commented:
> > I guess it would be preferable before merging these pipelines to think about a possible caching of gentoo image with dependencies and maybe create it new once in a month.
> 
> Maybe, but I have my reservations about random docker images, with having to deal with security questions, rate limiting, and whatnot. I think 15 minutes for a CI run is not terribly bad, especially that people are more than welcome to run CI locally (where they can utilize local caching of images with little if any effort). If this becomes too much of a hassle, we can reconsider later.

FWIW, you could do something like https://github.com/mesonbuild/meson/blob/master/.github/workflows/images.yml

https://github.com/mesonbuild/meson/tree/master/ci/ciimage

We install packages, then run all tests, and on passing, deploy a new docker image solely for testing. This both speeds up CI and ensures CI failure never happens just because of distro dependency updates, but indicates a real project defect. (Dependency updates can be investigated separately when weekly cron jobs fail, without blocking merges to `master`.)

Github provides its own docker image registry too, these days.
